### ISSUE TITLE
Using weigth_decay slows down Adam optimizer over time.

### ISSUE THREAD

**Issue by johannespitz** [pytorch/pytorch#51539]
## 🐛 Bug

Adding weight_decay to the Adam optimizer, via the keyword argument, causes training iterations to slow down over time. In the provided example I see a slowdown of 2x to 3x (compared to the first few iterations) within a couple of seconds, when running on a CPU.

## To Reproduce

Run the following snippet with `--weight-decay` and without. 
Note, `amsgrad=True` results in similar behavior, but other optimizers I have tested do not show the same effect.

```
import time
import torch
import torch.nn as nn
from torch.optim import Adam, SGD, AdamW, Adadelta


class MLPQFunction(nn.Module):
    def __init__(self, obs_dim, hidden_dim):
        super().__init__()
        self.q = nn.Sequential(
            nn.Linear(obs_dim, hidden_dim),
            nn.ReLU(),
            nn.Linear(hidden_dim, 1)
        )
    def forward(self, obs, act):
        return self.q(obs)

def sac(device, lr=3e-4, weight_decay=0.0):
    q_function = MLPQFunction(20, 256)
    q_function.to(device)
    q_optimizer = Adam(q_function.parameters(), lr=lr, weight_decay=weight_decay)
    start = time.perf_counter()
    last = start
    for j in range(1000000):
        q_optimizer.zero_grad()
        o, a = torch.rand(256, 20, device=device), torch.rand(256, 20, device=device)
        q = q_function(o,a)
        backup = torch.rand(256, device=device)
        loss_q = ((q - backup)**2).mean()
        loss_q.backward()
        q_optimizer.step()
        if j % 500 == 0:
            curr = time.perf_counter()
            print(f'Iteration: {j}, Time: {curr - start:.2f}s')
            print(f'Time since last iteration: {curr - last:.2f}s')
            last = curr


if __name__ == '__main__':
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument('--weight-decay', type=float, default=0.0)
    parser.add_argument('--cuda', action='store_true', default=False)
    args = parser.parse_args()
    use_cuda = args.cuda and torch.cuda.is_available()
    device = torch.device("cuda" if use_cuda else "cpu")
    if not use_cuda:
        torch.set_num_threads(2)

    try:
        sac(device, weight_decay=args.weight_decay)
    except KeyboardInterrupt:
        pass
```

## Expected behavior

```
python minimal.py                      
Iteration: 0, Time: 0.01s
Time since last iteration: 0.01s
Iteration: 500, Time: 0.42s
Time since last iteration: 0.41s
Iteration: 1000, Time: 0.83s
Time since last iteration: 0.41s
Iteration: 1500, Time: 1.25s
Time since last iteration: 0.41s
Iteration: 2000, Time: 1.65s
Time since last iteration: 0.41s
Iteration: 2500, Time: 2.07s
Time since last iteration: 0.42s
Iteration: 3000, Time: 2.48s
Time since last iteration: 0.41s
Iteration: 3500, Time: 2.89s
Time since last iteration: 0.41s
Iteration: 4000, Time: 3.30s
Time since last iteration: 0.41s
Iteration: 4500, Time: 3.71s
Time since last iteration: 0.41s
Iteration: 5000, Time: 4.12s
Time since last iteration: 0.41s
```

## Actual behavior

```
minimal.py --weight-decay 0.0025
Iteration: 0, Time: 0.01s
Time since last iteration: 0.01s
Iteration: 500, Time: 0.43s
Time since last iteration: 0.43s
Iteration: 1000, Time: 0.87s
Time since last iteration: 0.44s
Iteration: 1500, Time: 1.30s
Time since last iteration: 0.43s
Iteration: 2000, Time: 1.80s
Time since last iteration: 0.50s
Iteration: 2500, Time: 2.40s
Time since last iteration: 0.60s
Iteration: 3000, Time: 3.10s
Time since last iteration: 0.70s
Iteration: 3500, Time: 3.91s
Time since last iteration: 0.81s
Iteration: 4000, Time: 4.85s
Time since last iteration: 0.93s
Iteration: 4500, Time: 5.90s
Time since last iteration: 1.05s
Iteration: 5000, Time: 7.12s
Time since last iteration: 1.22s
```

## Environment

I encounter the same problem on different machines with different OS.
I install the lastest pytorch version via anacona.

```
python collect_env.py 
Collecting environment information...
PyTorch version: 1.7.1
Is debug build: False
CUDA used to build PyTorch: 11.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.10.2

Python version: 3.7 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce RTX 3070
Nvidia driver version: 460.32.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] torch==1.7.1
[pip3] torchaudio==0.7.0a0+a853dff
[pip3] torchvision==0.8.2
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               11.0.221             h6bb024c_0  
[conda] mkl                       2020.2                      256  
[conda] mkl-service               2.3.0            py37he8ac12f_0  
[conda] mkl_fft                   1.2.0            py37h23d657b_0  
[conda] mkl_random                1.1.1            py37h0573a6f_0  
[conda] numpy                     1.19.2           py37h54aff64_0  
[conda] numpy-base                1.19.2           py37hfa32c7d_0  
[conda] pytorch                   1.7.1           py3.7_cuda11.0.221_cudnn8.0.5_0    pytorch
[conda] torchaudio                0.7.2                      py37    pytorch
[conda] torchvision               0.8.2                py37_cu110    pytorch
```

cc @vincentqb

**Comment by johannespitz** [comment]
I think, maybe it is related to issue #20146.

**Comment by johannespitz** [comment]
The problem is still there. I tested it just now on a MacBook, with the latest nightly build (1.9.0.dev20210221).
@H-Huang

**Comment by coreylowman** [comment]
Running into same issue. Looking at the MR mentioned in #20146, it seems like that change isn't there in the source anymore?

**Comment by cijerezg** [comment]
I am running into the same issue.

**Comment by ldv1** [comment]
Two years later, I am experiencing the same issue with PyTorch 1.11.0.

**Comment by ldv1** [comment]
Using the above code with pytorch  '2.0.1+cpu', training over 50k iterations  runs in 57s  without weight decay. With a weight decay of 1e-5, it runs in 72s; with a weight decay of 1e-1 it runs in 186s  and with a weigh decay of 1e1 it runs in 240s.

**Comment by smanschi** [comment]
Hi,

I got the same issue with pytorch v2.0. As mentioned here already, there is a memory leak in the Adam optimizer (and other optimizers) when using weight decay. I checked the history of the Adam optimizer it seems like a fix has been added but was removed in v1.5.0.

v1.4.1: https://github.com/pytorch/pytorch/blob/v1.4.1/torch/csrc/api/src/optim/adam.cpp#L26 (with fix)
v1.5.0: https://github.com/pytorch/pytorch/blob/v1.5.0-rc1/torch/csrc/api/src/optim/adam.cpp#L108 (removed fix)
v2.0.1:https://github.com/pytorch/pytorch/blob/v2.0.1/torch/csrc/api/src/optim/adam.cpp#L121

Can anybody comment on why the fix has been removed?

**Comment by ldv1** [comment]
I do not believe it has to do with a memory leak in Adam.
If I replace the weight decay in Adam by the standard (and slow) implementation of L2 regularization:

`loss_q += weight_decay*sum(p.pow(2).sum() for p in q_function.parameters() if p.requires_grad)`

then I observe the same trend: for 50k iterations, training runs in ~80s without regularization , ~130s with a regularization factor of 1e-5, ~200s for a factor of 0.1 and ~400s if the factor is as high as 10. Runtime should be independent of the weight decay factor. Do I err?

**Comment by smanschi** [comment]
Hmm.. did you check if the training time per epoch increases over time? I also find it strange that for me the training time increased after a while until it plateaus again. For a "classic" memory leak I would expect a constant increase over time.

**Comment by johannespitz** [comment]
I am not sure anymore where I read that, but this behavior is probably due to the operations becoming less and less sparse over time. (Since weight decay encourages using many small weights rather than a fewer bigger ones.)
So this likely not a bug, just a curios consequence of the highly optimized sparse matrix multiplications.
Try using L1 norm if you need faster iterations.

**Comment by ldv1** [comment]
Though this issue is closed, I want to share what worked for me:

`torch.set_flush_denormal(True)`

As stated in https://discuss.pytorch.org/t/training-time-gets-slower-and-slower-on-cpu/145483/5 for another experiment,  

> numbers gets very close to 0. Your CPU, to try and keep precision, has to do some fancy arithmetic and thus gets slowed down

This is also true with large  weight decays.
